File: Algorithms/leetcode/Ex_404_Sum_of_Left_Leaves/Soln_3_Recursive_Dfs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def sumOfLeftLeaves(self, root: TreeNode) -> int:
        return self.dfs(root, 0)

    def dfs(self, node: TreeNode, leftSum: int) -> int:
        if not node:
            return leftSum

        incre = 0
        if node.left:
            if not node.left.left and not node.left.right:
                incre += node.left.val
            else:
                incre += self.dfs(node.left, leftSum)
        if node.right:
            incre += self.dfs(node.right, leftSum)
        return leftSum + incre

# 注：
# 这个解法相较于解法2，减少了一个参数，每次计算增量，然后返回当前的leftSum加上增量后的结果

File: Algorithms/leetcode/Ex_0297_Serialize_and_Deserialize_Binary_Tree/Soln_3_Recursive_Dfs.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
var serialize = function(root) {
    if (root == null) {
        return 'X';
    }
    const left = serialize(root.left);
    const right = serialize(root.right);
    return root.val + ',' + left + ',' + right; // 顺序：根 左 右
};

/**
 * Decodes your encoded data to tree.
 *
 * @param {string} data
 * @return {TreeNode}
 */
var deserialize = function(data) {
    const list = data.split(',');

    const buildTree = (list) => {
        const rootVal = list.shift();   // 弹出首项
        if (rootVal == 'X') {
            return null;
        }
        const root = new TreeNode(rootVal); // 不是X，创建节点
        root.left = buildTree(list);    // 递归构建左子树
        root.right = buildTree(list);   // 递归构建右子树
        return root;
    }

    return buildTree(list);
};

/**
 * Your functions will be called as such:
 * deserialize(serialize(root));
 */

// 参考题解：https://leetcode-cn.com/problems/serialize-and-deserialize-binary-tree/solution/shou-hui-tu-jie-gei-chu-dfshe-bfsliang-chong-jie-f/

